skbio/metadata/_interval.py

```diff
@@ -0,0 +1,456 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from ._intersection import IntervalTree
+from skbio.util._decorator import experimental
+from itertools import chain
+
+
+class Interval:
+    '''Stores the position and metadata of an interval.
+
+    This class supports the storing data for contiguous intervals
+    and non-contiguous intervals.
+
+    Parameters
+    ----------
+    interval_metadata : object
+        A reference to the `IntervalMetadata` object that this
+        interval is associated to.
+    intervals : iterable of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : iterable of tuple of bool
+        List of tuples, representing the openness of each interval.
+        If this isn't specified, then all of the boundaries are true.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+
+    See Also
+    --------
+    skbio.metadata.IntervalMetadata
+    '''
+    def __init__(self, interval_metadata, intervals,
+                 boundaries=None, metadata=None):
+
+        ivs = list(intervals)
+
+        # Intervals
+        self._interval_metadata = interval_metadata
+        # http://stackoverflow.com/a/6422754/1167475
+        # Used to sort boundaries later
+        indices = [i[0] for i in sorted(enumerate(ivs),
+                                        key=lambda x:x[1])]
+        self.intervals = sorted(ivs)
+
+        # Boundaries
+        if boundaries is not None:
+            bds = list(boundaries)
+            if len(bds) != len(ivs):
+                msg = ('The number of boundaries (%d) needs '
+                       'to be equal to the number of intervals (%d).')
+                raise ValueError(msg % (len(bds), len(ivs)))
+            self.boundaries = [bds[i] for i in indices]
```

Comment:
oh, I see that now. Not sure if it makes sense, but if `ivs` and `boundaries` were a `np.array` then some operations like this become more straight forward to express w/o type conversion. (Just a comment, can defer to later and non-blocking, may have negative implications elsewhere). For instance, the sorting on line 47 can be done with `np.argsort` which then allowing for `self.boundaries` to be sorted with fancy indexing.